arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,307 @@
+  calculate_aligned_edges( const vil_image_view< PixType >& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j );
+
+  template < typename InputType,
+             typename OutputType > vil_image_view< OutputType >
+  nonmax_suppression( const vil_image_view< InputType >& grad_i,
+                      const vil_image_view< InputType >& grad_j );
+
+  template < typename pix_t > vil_image_view< pix_t >
+  filter( const vil_image_view< pix_t >& input_image );
+
+  aligned_edge_detection* p;
+  // Internal parameters/settings
+  float threshold = 10;
+  bool produce_joint_output = true;
+  double smoothing_sigma = 1.3;
+  unsigned smoothing_half_step = 2;
+};
+
+// Perform NMS on the input gradient images in horizontal and vert directions
+// only
+template < typename InputType, typename OutputType >
+vil_image_view< OutputType >
+aligned_edge_detection::priv
+::nonmax_suppression( const vil_image_view< InputType >& grad_i,
+                      const vil_image_view< InputType >& grad_j )
+{
+  if( grad_i.ni() != grad_j.ni() || grad_i.nj() != grad_j.nj() )
+  {
+    LOG_ERROR(
+      p->logger(),
+      "Input gradient image dimensions must be equivalent" );
+  }
+
+  const unsigned ni = grad_i.ni();
+  const unsigned nj = grad_i.nj();
+
+  vil_image_view< OutputType > output( ni, nj, 2 );
+  output.fill( 0 );
+
+  // Perform non-maximum suppression
+  for( unsigned j = 1; j < nj - 1; j++ )
+  {
+    for( unsigned i = 1; i < ni - 1; i++ )
+    {
+      const InputType val_i = grad_i( i, j );
+      const InputType val_j = grad_j( i, j );
+
+      if( val_i > threshold )
+      {
+        if( val_i >= grad_i( i - 1, j ) && val_i >= grad_i( i + 1, j ) )
+        {
+          output( i, j, 0 ) = static_cast< OutputType >( val_i );
+        }
+      }
+      if( val_j > threshold )
+      {
+        if( val_j >= grad_j( i, j - 1 ) && val_j >= grad_j( i, j + 1 ) )
+        {
+          output( i, j, 1 ) = static_cast< OutputType >( val_j );
+        }
+      }
+    }
+  }
+  return output;
+}
+
+// Calculate potential edges
+template < typename PixType, typename GradientType >
+vil_image_view< PixType >
+aligned_edge_detection::priv
+::calculate_aligned_edges( const vil_image_view< PixType >& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j )
+{
+  // Calculate sobel approx in x/y directions
+  vil_sobel_3x3( input, grad_i, grad_j );
+
+  // Take absolute value of gradients
+  vil_transform< GradientType, GradientType( GradientType ) >( grad_i,
+                                                               std::abs );
+  vil_transform< GradientType, GradientType( GradientType ) >( grad_j,
+                                                               std::abs );
+
+  // Perform NMS in vert/hori directions and threshold magnitude
```

Comment:
Two of these long lines below have rather arbitrary line breaks. Try breaking after the equal to see if we can make the code more readable. Breaking early should allow less line breaks.

Reply:
Thanks for the catch. This formatting was induced by `uncrustify` so if I get a chance I'll see if there's a config option to get it to break earlier. However, the current config did not regress the formatting when I manually broke at the equals.

Reply:
I believe when breaks are added, they are added as late as possible. The preferred solution is to manually add breaks at more sensible locations. (In my experience, it's effectively impossible to write tools that are as clever as a human. As I've noted elsewhere, *I* view the goal of tool assisted formatting not as "make it perfect", but "accept correct formatting, and let you know if it's wrong". By contrast, clang-format might *enforce* the poor formatting in a case like this...)